### ISSUE TITLE
Provisioning of GKE-Cluster not working (configured proxy)

### ISSUE THREAD

**Issue by ChrisHaPunkt** [rancher/rancher#22006]
**What kind of request is this (question/bug/enhancement/feature request):**
BUG

**Steps to reproduce (least amount of steps as possible):**
Don't know if this is related to our proxy

1. Configure a http-proxy in Rancher Helm-Chart 
a. proxy: http://techproxy.<domain\>.local:80
b. noProxy: 127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16
2. Generate a valid ServiceAccount in the GCP
3. Create GKE Cluster through UI (we disabled stackdriver monitoring and logging)

**Result:**
- "Provisioning cluster c-hb67b....."
- Cluster is up in GKE
![grafik](https://user-images.githubusercontent.com/4389395/62524083-51ffb200-b835-11e9-98d9-0d34e74d8330.png)
- "Waiting for API to be available"
- No cluster and node agent ressources are deployed in the GKE Cluster
- "Waiting for API to be available" infinitely
Is it prossible that the proxy is not applied to some sub-container during cluster creation or sth?
**Other details that may be helpful:**

- No Rancher cluster and node agent ressources are deployed to the GKE Cluster
![grafik](https://user-images.githubusercontent.com/4389395/62525283-6e9ce980-b837-11e9-8b74-dee7feb63abd.png)

- The endpoint url (from the logs below) can be curled and accessed from within the rancher pod. The proxy vars are set there correctly.
- The GKE Cluster is being deleted if I trigger the deletion in Rancher 

**Rancher logs flooding with the following**
> 19/08/06 08:31:42 [INFO] Stopping cluster agent for c-hb67b
> 2019/08/06 08:31:42 [ERROR] ClusterController c-z8rqx [user-controllers-controller] failed with : failed to start user controllers for cluster c-hb67b: failed to contact server: Get https://<gke-endpoint\>/api/v1/namespaces/kube-system?timeout=30s: proxyconnect tcp: waiting for cluster agent to connect
> 2019/08/06 08:31:43 [INFO] Stopping cluster agent for c-hb67b
> 2019/08/06 08:31:43 [ERROR] ClusterController c-vnrwk [user-controllers-controller] failed with : failed to start user controllers for cluster c-hb67b: failed to contact server: Get https://<gke-endpoint\>/api/v1/namespaces/kube-system?timeout=30s: proxyconnect tcp: waiting for cluster agent to connect
> 2019/08/06 08:31:43 [INFO] Stopping cluster agent for c-hb67b
> 2019/08/06 08:31:43 [ERROR] ClusterController c-vnrwk [user-controllers-controller] failed with : failed to start user controllers for cluster c-hb67b: failed to contact server: Get https://<gke-endpoint\>/api/v1/namespaces/kube-system?timeout=30s: proxyconnect tcp: waiting for cluster agent to connect
> ...

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.6
- Installation option (single install/HA): HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Hosted GKE
- Machine type (cloud/VM/metal) and specifications (CPU/memory): cloud
- Kubernetes version (use `kubectl version`): 1.13.7

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:46:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.7-gke.8", GitCommit:"7d3d6f113e933ed1b44b78dff4baf649258415e5", GitTreeState:"clean", BuildDate:"2019-06-19T16:37:16Z", GoVersion:"go1.11.5b4", Compiler:"gc", Platform:"linu
x/amd64"}
```

- Docker version (use `docker version`):

```
18.09.6
```

**Comment by janeczku** [comment]
Reproduced the issue in v2.2.8 following the steps above but provisioning an EKS cluster instead.

The only logged error is:

> 2019/09/19 12:27:22 [ERROR] ClusterController c-9nc2j [user-controllers-controller] failed with : failed to start user controllers for cluster c-9nc2j: failed to contact server: Get https://REDACTED.gr7.eu-central-1.eks.amazonaws.com/api/v1/namespaces/kube-system?timeout=30s: proxyconnect tcp: waiting for cluster agent to connect

The EKS Kubernetes cluster does have the `cattle-system` namespace, but it is empty.

This all points to the the issue occurring during generating and deploying the Rancher agent YAML.
I was able to deploy the Rancher agent YAML manually, which caused the cluster to become healthy.

**Comment by tuxmonteiro** [comment]
+1

**Comment by vintbra** [comment]
+1

**Comment by alena1108** [comment]
@tuxmonteiro @ChrisHaPunkt do you have `localhost,127.0.0.1' added to NO_PROXY? Rancher uses localhost connection to run kubectl apply to deploy cluster-agent https://github.com/rancher/rancher/blob/f139213011c9cdffd0baecaf1ac26c9fb85ec2b9/pkg/clustermanager/manager.go#L462 , and I wonder if that request tries to go through proxy and fails.

**Comment by alena1108** [comment]
The root cause is: cluster gets stuck at the user controller registration point. Cluster/node agent kubectl deployment gets called after user controllers are registered, and that never happens.

The reason for the user controller registration failure is in the way we build the dialer to the hosted k8s clusters. All the requests should go directly to hosted k8s api endpoint, but this condition will never be true if proxy environment is set: 

https://github.com/rancher/rancher/blob/master/pkg/dialer/factory.go#L139
`address` will be a proxy ip address, and `hostPort` is the k8s api endpoints, and they'll never match

**Comment by jiaqiluo** [comment]
The bug is reproduced in Rancher:v2.3.3

Steps:
- run Rancher behind a proxy
- add a GKE cluster from Rancher's UI

Results:
- the cluster is created and active in Google's console
- the namespace `cattle-system` is created in the cluster with no pods in it
- The cluster is stuck in `Waiting for API to be available` in Rancher
- Rancher logs shows the error:
```
2019/12/09 22:16:25 [ERROR] ClusterController _all_ [user-controllers-controller] 
failed with : failed to start user controllers for cluster c-whvph: failed to 
contact server: Get https://35.190.154.236/api/v1/namespaces/kube-systemtimeout=30s: proxyconnect tcp:
waiting for cluster agent to connect
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in master-head ed99b31be

Steps:
- run Rancher behind a proxy
- add a GKE cluster from Rancher's UI

Results:
- the cluster is active in Rancher
- can deploy workloads
- can access to the kubectl via UI

**Comment by jiaqiluo** [comment]
The bug fix is validated in v2.3-head 7195064f9

Steps:
- run Rancher behind a proxy
- add a GKE cluster from Rancher's UI

Results:
- the cluster is active in Rancher
- no error in Rancher log
- can deploy workloads
- can access to the kubectl via UI

**Comment by vaskoevgen** [comment]
Hello everyone. I have the same issue with Rancher for GKE. How can I setup rancher again from scratch. I had tried delete and add again but I had no successful result.
![image](https://user-images.githubusercontent.com/20374729/71772521-fa38b480-2f54-11ea-9aa5-87132ad19251.png)
